rework section on provider/repository (#468)
jakartaee/data: spec/src/main/asciidoc/jakarta-ee.adoc

== Jakarta Data Providers

A Jakarta Data provider might come as an integrated component of a Jakarta EE container, or it might be a separate component that integrates with the Jakarta EE container via standard or proprietary SPIs.

> For example, a Jakarta Data provider might use a CDI portable extension to integrate with dependency injection.

Reviewer: ```suggestion
A Jakarta Data provider can use a CDI portable extension to integrate with dependency injection.
```

Author: Not sure about this change. I was trying to be illustrative here, not prescriptive.

Reviewer: Sure, that's fine. I'm pretty sure we state this elsewhere, so we can leave this as you wrote it and it will be fine.  I'll mark this comment resovled.

The role of the Jakarta Data provider is to supply implementations of application-defined repositories.

A Jakarta Data provider must supply an implementation of each repository interface which either:

- explicitly specifies the name of the provider via the `provider` member of the `@Repository` annotation, or
- does not explicitly specify a `provider`, but whose entity classes are annotated with an entity-defining annotation type supported by the provider.

> The repository implementation, and the provider itself, is responsible for mediating interaction between the application program and the datastore, often taking advantage of other Jakarta EE services such as datasource provisioning and transaction management.

Reviewer: ```suggestion
The repository implementation and the Jakarta Data provider are responsible for mediating interaction between the application program and the datastore, often taking advantage of other Jakarta EE services such as datasource provisioning and transaction management.
```

Author: Applied the change.

An application usually obtains a reference to an implementation of a repository via dependency injection.
Therefore, the provider must make its repository implementations available to the bean container responsible for dependency injection.

> Multiple Jakarta Data provider may coexist within a single program, and provide repository implementations.

Reviewer: ```suggestion
Multiple Jakarta Data providers may coexist within a single program and provide repository implementations.
```

Author: Applied the change.

It is the responsibility of the application developer to ensure that each repository is eligible for implementation by exactly one provider, according to the rules specified below.

=== Provider support for Entities

> A Jakarta Data provider supports one or more entity-defining annotation types.

Reviewer: This is another subtle change that inadvertently rules out providers supporting unannotated entities.  We can adjust it as follows to preserve the meaning:
```suggestion
A Jakarta Data provider typically supports one entity-defining annotation type, but can support any number of entity-defining annotation types.
```

Author: You really want to allow providers which support _no_ entity-defining annotation?

OK, I guess.

Reviewer: I remember it being discussed a while back, and wanted to be sure we weren't accidentally removing it.  I did do some experimenting with the idea myself and it did seem workable, albeit limiting.

> For example:

Reviewer: The language "For example" implies the following paragraphs are examples and not requirements.
```suggestion
In the case of Jakarta Data providers that are built on Jakarta Persistence and Jakarta NoSQL:
```

Author: I changed it to "In particular:"

> - The `jakarta.persistence.Entity` annotation from the Jakarta Persistence specification is the entity-defining annotation for Jakarta Data providers backed by a Jakarta Persistence provider. Other Jakarta Data providers must not support the use of `jakarta.persistence.Entity` as an entity-defining annotation.

Reviewer: To allow for providers that support unannotated entities:
```suggestion
- The `jakarta.persistence.Entity` annotation from the Jakarta Persistence specification is an entity-defining annotation for Jakarta Data providers backed by a Jakarta Persistence provider. Other Jakarta Data providers must not support the use of `jakarta.persistence.Entity` as an entity-defining annotation.
```

> - The `jakarta.nosql.Entity` annotation from the Jakarta NoSQL specification is the entity-defining annotation for Jakarta Data providers backed by NoSQL databases. Other Jakarta Data providers must not support the use of `jakarta.nosql.Entity` as an entity-defining annotation.

Reviewer: To allow for providers that support unannotated entities:
```suggestion
- The `jakarta.nosql.Entity` annotation from the Jakarta NoSQL specification is an entity-defining annotation for Jakarta Data providers backed by NoSQL databases. Other Jakarta Data providers must not support the use of `jakarta.nosql.Entity` as an entity-defining annotation.
```

Author: Applied the change.

A Jakarta Data provider might define a custom entity-defining annotation.
The name of such a custom entity-defining annotation type must end with `Entity`. This convention allows any Jakarta Data provider to identify when a repository is associated with entity classes declared using an unrecognized entity-defining annotation.

> A Jakarta Data provider must ignore every repository associated with entity classes declared using an unrecognized entity-defining annotation, allowing another Jakarta Data provider to supply the implementation.

Reviewer: There are some subtle, and probably unintentional, changes in meaning here.  The previous language allows for the scenario where an entity class is compiled with multiple entity-defining annotations but only one is available at run time (The TCK actually does this so it can run against either type of provider).  The previous language also allows for a provider that supports a simpler model with unannotated entities.  I believe we can adjust your rewrite as follows to preserve the meaning:
```suggestion
A Jakarta Data provider must ignore every repository associated with entity classes where entity-defining annotations are available at run time but none are recognized by the provider, allowing another Jakarta Data provider to supply the implementation.
```

Author: I see. So that scenario wasn't obvious. I would add language explicitly talking about this usage pattern (unless it's already somewhere else in the spec that I missed).

Author: I have added language making this more explicit, WDYT, Nathan?
